src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -148,6 +154,44 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                                     // Non-public properties should not be included for (de)serialization.
                                 }
                             }
+
+                            foreach (FieldInfo fieldInfo in currentType.GetFields(binding))
```

Comment:
We need to extend the property visibility tests added in https://github.com/dotnet/runtime/pull/32107 and https://github.com/dotnet/runtime/pull/36936 to include collisions between field names, and collisions between property and field names.